pyquil/api/_quantum_computer.py

```diff
@@ -669,45 +670,116 @@ def get_qc(name: str, *, as_qvm: bool = None, noisy: bool = None,
+
+    :returns: ``True`` if a process is listening on the specified host/port, ``False`` otherwise
+    """
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    try:
+        s.connect((host, port))
+        return True
+    except ConnectionRefusedError:
+        return False
+    finally:
+        s.close()
+
+
+@contextmanager
+def local_forest_runtime(
+        *,
+        host: str = '127.0.0.1',
+        qvm_port: int = 5000,
+        quilc_port: int = 5555,
+        use_protoquil: bool = False) -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
+    """A context manager for local QVM and QUIL compiler.
+
     You must first have installed the `qvm` and `quilc` executables from
     the forest SDK. [https://www.rigetti.com/forest]
 
-    This context manager will start up external processes for both the
-    compiler and virtual machine, and then terminate them when the context
-    is exited.
+    This context manager will ensure that the designated ports are not used, start up `qvm` and
+    `quilc` proccesses if possible and terminate them when the context is exited.
+    The returned tuple contains two ``subprocess.Popen`` objects for the `qvm` and the `quilc` processes.
+    If one of the ports is in use, a ``RuntimeWarning`` will be issued, the `qvm`/`quilc` process
+    won't be started and the respective value in the tuple will be ``None``.
+
+    .. note::
+        Only processes started by this context manager will be terminated on exit, no external process will
+        be touched.
 
-    If `qvm` (or `quilc`) is already running, then the existing process will
-    be used, and will not terminated at exit.
 
     >>> from pyquil import get_qc, Program
     >>> from pyquil.gates import CNOT, Z
-    >>> from pyquil.api import local_qvm
+    >>> from pyquil.api import local_forest_runtime
     >>>
     >>> qvm = get_qc('9q-square-qvm')
     >>> prog = Program(Z(0), CNOT(0, 1))
     >>>
-    >>> with local_qvm():
+    >>> with local_forest_runtime():
     >>>     results = qvm.run_and_measure(prog, trials=10)
 
+    :param host: Host on which `qvm` and `quilc` should listen on.
+    :param qvm_port: Port which should be used by `qvm`.
+    :param quilc_port: Port which should be used by `quilc`.
+    :param use_protoquil: Restrict input/output to protoquil.
+
+    .. warning::
+        If ``use_protoquil`` is set to ``True`` language features you need
+        may be disabled. Please use it with caution.
+
     :raises: FileNotFoundError: If either executable is not installed.
     """
-    # Enter. Acquire resource
-    qvm = subprocess.Popen(['qvm', '-S'],
-                           stdout=subprocess.PIPE,
-                           stderr=subprocess.PIPE)
 
-    quilc = subprocess.Popen(['quilc', '-RP'],
-                             stdout=subprocess.PIPE,
-                             stderr=subprocess.PIPE)
+    qvm = None
+    quilc = None
+
+    # if host is 0.0.0.0 replace it with 127.0.01
```

Comment:
Rather than documenting *what* is happening (the code does that already), document *why*.